ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Setup.hs

```diff
@@ -56,24 +59,57 @@ runTest schedulerConfig genesisTest schedule makeProperty = do
       "    mustReply = " ++ show (mustReplyTimeout scChainSyncTimeouts)
       ] ++ prettyGenesisTest genesisTest
 
-    finalStateView <- runPointSchedule schedulerConfig genesisTest schedule tracer
-    traceWith tracer (condense finalStateView)
-    trace <- unlines <$> getTrace
+    rgtrStateView <- runPointSchedule schedulerConfig genesisTest schedule tracer
+    traceWith tracer (condense rgtrStateView)
+    rgtrTrace <- unlines <$> getTrace
 
-    pure $ counterexample trace $ makeProperty finalStateView
+    pure $ RunGenesisTestResult {rgtrTrace, rgtrStateView}
   where
     SchedulerConfig {scChainSyncTimeouts} = schedulerConfig
     GenesisTest {gtBlockTree} = genesisTest
 
--- | Print counterexamples if the test result contains exceptions.
-exceptionCounterexample :: Testable a => (StateView -> [PeerId] -> a) -> StateView -> Property
-exceptionCounterexample makeProperty stateView =
-  case svChainSyncExceptions stateView of
-    exns | ([], killed) <- partitionEithers (genesisException <$> exns) ->
-      property $ makeProperty stateView killed
-    exns ->
-      counterexample ("exceptions: " <> show exns) False
+-- | Variant of 'runGenesisTest' that also takes a property on the final
+-- 'StateView' and returns a QuickCheck property. The trace is printed in case
+-- of counter-example.
+runGenesisTest' ::
+  Testable prop =>
+  SchedulerConfig ->
+  GenesisTest ->
+  PointSchedule ->
+  (StateView -> prop) ->
+  Property
+runGenesisTest' schedulerConfig genesisTest schedule makeProperty =
+    counterexample rgtrTrace $ makeProperty rgtrStateView
   where
-    genesisException = \case
-      (ChainSyncException peer e) | Just ThreadKilled <- fromException e -> Right peer
-      exc -> Left exc
+    RunGenesisTestResult{rgtrTrace, rgtrStateView} =
+      runGenesisTest schedulerConfig genesisTest schedule
+
+type ForAllGenesisTest schedule prop =
+  Gen (GenesisTest, schedule) ->
+  SchedulerConfig ->
+  (GenesisTest -> schedule -> StateView -> prop) ->
+  Property
```

Comment:
I would like to add the `Testable prop =>` in there and not require `prop` as an argument of `ForAllGenesisTest`, so as to simplify the type signature of the following functions. I'm sure there is a language extension for this?

Reply:
Maybe inlining the type synonym at the few places where it is used would offer the less friction.

Reply:
You mean 072962b58458436a0b6662a1c04131ad227ed578?